cmd/run.go

```diff
@@ -288,71 +288,20 @@ func Run(isDebug *bool) *cli.Command {
 				return cli.Exit("", 1)
 			}
 
-			runMain := true
-			runChecks := true
-			runPushMetadata := c.Bool("push-metadata") || foundPipeline.MetadataPush.HasAnyEnabled()
-			if runPushMetadata {
-				foundPipeline.MetadataPush.Global = true
-			}
-
-			onlyFlags := c.StringSlice("only")
-			if len(onlyFlags) > 0 {
-				runMain = slices.Contains(onlyFlags, "main")
-				runChecks = slices.Contains(onlyFlags, "checks")
-				runPushMetadata = slices.Contains(onlyFlags, "push-metadata")
-
-				for _, flag := range onlyFlags {
-					if flag != "main" && flag != "checks" && flag != "push-metadata" {
-						errorPrinter.Printf("Invalid value for '--only' flag: '%s', available values are 'main', 'checks', and 'push-metadata'\n", flag)
-						return cli.Exit("", 1)
-					}
-				}
+			filter := &Filter{
+				IncludeTag:        c.String("tag"),
+				OnlyTaskTypes:     c.StringSlice("only"),
+				IncludeDownstream: runDownstreamTasks,
+				PushMetaData:      c.Bool("push-metadata"),
+				SingleTask:        task,
 			}
 
 			s := scheduler.NewScheduler(logger, foundPipeline)
-
-			// mark all the instances to be skipped, then conditionally mark the ones to run to be pending
-			s.MarkAll(scheduler.Pending)
-			if task != nil {
-				logger.Debug("marking single task to run: ", task.Name)
-				s.MarkAll(scheduler.Succeeded)
-				s.MarkAsset(task, scheduler.Pending, runDownstreamTasks)
-
-				if c.String("tag") != "" {
-					errorPrinter.Printf("You cannot use the '--tag' flag when running a single asset.\n")
-					return cli.Exit("", 1)
-				}
-			}
-
 			sendTelemetry(s, c)
```

Comment:
Here we are sending the telemetry that gives us info on the assets that are going to be run *before* knowing what assets are going to be run. I think this needs to be moved after the filtering and marking